src/virtualdevices/dbus/dbus.rs

```diff
@@ -0,0 +1,348 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::io;
+use std::sync::mpsc::channel;
+use std::sync::{Arc, Mutex};
+use zbus::{dbus_proxy, Result};
+use zvariant::{OwnedObjectPath, OwnedValue, Value};
+
+use crate::errors;
+use crate::{RegisterResult, SignResult};
+
+#[dbus_proxy(
+    interface = "org.freedesktop.fido2.Device",
+    default_service = "org.freedesktop.fido2"
+)]
+trait Device {
+    fn make_credential(
+        &self,
+        sign_type: &str,
+        client_data_hash: &[u8],
+        rp: &str,
+        user_id: &[u8],
+        user: &str,
+        options: HashMap<&str, &Value>,
+    ) -> Result<OwnedObjectPath>;
+
+    fn get_assertion(
+        &self,
+        client_data_hash: &[u8],
+        rp: &str,
+        key_handle: &[u8],
+        options: HashMap<&str, &Value>,
+    ) -> Result<OwnedObjectPath>;
+}
+
+#[dbus_proxy(
+    interface = "org.freedesktop.fido2.Request",
+    default_service = "org.freedesktop.fido2"
+)]
+trait Request {
+    fn cancel(&self) -> Result<()>;
+
+    #[dbus_proxy(signal)]
+    fn error(&self) -> Result<(u32, String)>;
+
+    #[dbus_proxy(signal)]
+    fn completed(&self) -> Result<HashMap<String, OwnedValue>>;
+}
+
+pub struct DeviceManagerState<'c> {
+    pub devices: Vec<DeviceProxy<'c>>,
+}
+
+impl<'c> DeviceManagerState<'c> {
+    pub fn new() -> Arc<Mutex<DeviceManagerState<'c>>> {
+        Arc::new(Mutex::new(DeviceManagerState { devices: vec![] }))
+    }
+}
+
+pub fn serve(
+    state: Arc<Mutex<DeviceManagerState<'static>>>,
+    connection: zbus::Connection,
+) -> Result<()> {
+    let object_manager_connection = zbus::Connection::new_session()?;
+
+    let object_manager = zbus::fdo::ObjectManagerProxy::new_for(
+        &object_manager_connection,
+        "org.freedesktop.fido2",
+        "/org/freedesktop/fido2/Device",
+    )?;
+
+    let connectionclone = connection.clone();
+    let stateclone = state.clone();
+
+    object_manager
+        .inner()
+        .connect_signal("InterfacesAdded", move |message| {
+            let (object_path, _): (
+                zvariant::OwnedObjectPath,
+                HashMap<&str, HashMap<&str, zvariant::Value>>,
+            ) = message.body()?;
+            let device = DeviceProxy::new_for_owned_path(
+                connectionclone.clone(),
+                object_path.as_str().to_string(),
+            )?;
+            let mut lock = stateclone.lock().unwrap();
+            lock.devices.push(device);
+            Ok(())
+        })?;
+
+    let stateclone = state.clone();
+
+    object_manager
+        .inner()
+        .connect_signal("InterfacesRemoved", move |message| {
```

Comment:
Oh, zbus should provide a more specific `connect_interfaces_removed_signal` here. It'd be actually super trivial to add so I'll do that today. I'll check if other signals are missing from fdo interfaces as well and add them.

Reply:
You can track that [here](https://gitlab.freedesktop.org/zeenix/zbus/-/issues/119).

Reply:
@ueno [Here you go](https://gitlab.freedesktop.org/zeenix/zbus/-/merge_requests/212). If you could test it before I merge, that would be great.

Reply:
Yes, I confirm that it works with [this change](https://github.com/mozilla/authenticator-rs/files/5777088/fdo-signals.diff.gz).

Reply:
Cool, btw you also don't need `.inner()` call since `Deref<Target=zbus::Proxy>` is implemented.

Reply:
zbus 1.8.0 just released has this new API now.

Reply:
Thanks; updated to use it.